scripts/build_osm_network.py

```diff
@@ -0,0 +1,1051 @@
+    - path: str
+        The file path where the CSV file will be saved.
+    - **kwargs: keyword arguments
+        Additional arguments to be passed to the `to_csv` function of pandas.
+
+    Returns:
+    - None
+
+    If the DataFrame is not empty or does not have empty columns, it will be
+    written to the CSV file with NA values replaced by the first value in the
+    `NA_VALUES` list. If the DataFrame is empty or has empty columns, an empty
+    file will be created at the specified path.
+    """
+    if "na_rep" in kwargs:
+        del kwargs["na_rep"]
+    if not df.empty or not df.columns.empty:
+        return df.to_csv(path, **kwargs, na_rep=NA_VALUES[0])
+    else:
+        with open(path, "w") as fp:
+            pass
+
+
+def line_endings_to_bus_conversion(lines):
+    """
+    Converts line endings to bus connections.
+
+    This function takes a df of lines and converts the line endings to bus
+    connections. It performs the necessary operations to ensure that the line
+    endings are properly connected to the buses in the network.
+
+    Parameters:
+    lines (DataFrame)
+
+    Returns:
+    lines (DataFrame)
+    """
+    # Assign to every line a start and end point
+
+    lines["bounds"] = lines["geometry"].boundary  # create start and end point
+
+    lines["bus_0_coors"] = lines["bounds"].map(lambda p: p.geoms[0])
+    lines["bus_1_coors"] = lines["bounds"].map(lambda p: p.geoms[-1])
+
+    # splits into coordinates
+    lines["bus0_lon"] = lines["bus_0_coors"].x
+    lines["bus0_lat"] = lines["bus_0_coors"].y
+    lines["bus1_lon"] = lines["bus_1_coors"].x
+    lines["bus1_lat"] = lines["bus_1_coors"].y
+
+    return lines
+
+
+# tol in m
+def set_substations_ids(buses, distance_crs, tol=5000):
+    """
+    Function to set substations ids to buses, accounting for location
+    tolerance.
+
+    The algorithm is as follows:
+
+    1. initialize all substation ids to -1
+    2. if the current substation has been already visited [substation_id < 0], then skip the calculation
+    3. otherwise:
+        1. identify the substations within the specified tolerance (tol)
+        2. when all the substations in tolerance have substation_id < 0, then specify a new substation_id
+        3. otherwise, if one of the substation in tolerance has a substation_id >= 0, then set that substation_id to all the others;
+           in case of multiple substations with substation_ids >= 0, the first value is picked for all
+    """
+
+    buses["station_id"] = -1
+
+    # create temporary series to execute distance calculations using m as reference distances
+    temp_bus_geom = buses.geometry.to_crs(distance_crs)
+
+    # set tqdm options for substation ids
+    tqdm_kwargs_substation_ids = dict(
+        ascii=False,
+        unit=" buses",
+        total=buses.shape[0],
+        desc="Set substation ids ",
+    )
+
+    station_id = 0
+    for i, row in tqdm(buses.iterrows(), **tqdm_kwargs_substation_ids):
```

Comment:
If this loop takes some time - as the progress bar seems to indicate - it might be suitable for parallel processing, cf.:

https://github.com/PyPSA/pypsa-eur/blob/0470f119cf34cc0b3d1cc17559a0f719f6b0e1ee/scripts/build_industrial_production_per_country.py#L276-L277

Reply:
open todo

Reply:
Not required, given current speed of rule.